File: Problem_1 Level_Order_traversal.py
# BFS Solution:
# // Time Complexity : O(n)
# // Space Complexity : O(n)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No
#
# // Your code here along with comments explaining your approach

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        result = []
        if root == None:
            return result
        q = deque()
        q.append(root)
        while q:
            size = len(q)
            li = []
            for i in range(size):
                curr = q.popleft()
                li.append(curr.val)
                if curr.left != None:
                    q.append(curr.left)
                if curr.right != None:
                    q.append(curr.right)

            result.append(li)
        return result
# ***********************************************************************
# DFS Solution:
# // Time Complexity : O(n)
# // Space Complexity : O(n)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No
#
#
# // Your code here along with comments explaining your approach
# Definition for a binary tree node.

class Solution:
    def __init__(self):
        self.result = []

    def levelOrder(self, root):
        if root == None: return []
        self.dfs(root, 0)
        return self.result

    def dfs(self, root, level):
        # base
        if root == None: return

        # logic
        if level == len(self.result):
            self.result.append([])

        self.result[level].append(root.val)

        self.dfs(root.left, level + 1)
        self.dfs(root.right, level + 1)

File: Problem1_LevelOrderTraversal.java
// Time Complexity : o(n)
// Space Complexity : o(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


class Solution {
    //BFS using Queue 
    public List<List<Integer>> levelOrder(TreeNode root) {
        
        if(root==null)
            return new ArrayList<>();
        
        List<List<Integer>> result=new ArrayList<>();
        
        Queue<TreeNode> queue=new LinkedList<>();
          queue.offer(root);
          
        
        while(!queue.isEmpty())
        {
            int queueSize=queue.size();
            List<Integer> currLevel = new ArrayList<>();
            for(int i=0;i<queueSize;i++)
            {
                TreeNode currNode=queue.poll();
                currLevel.add(currNode.val);
                if(currNode.left!=null)
                    queue.add(currNode.left);
                if(currNode.right!=null)
                    queue.add(currNode.right);
            }
            result.add(currLevel);
        }
        return result;
    }
    /*
    // using dfs
    public List<List<Integer>> levelOrder(TreeNode root) {
        
        if(root==null)
            return new ArrayList<>();
        
        List<List<Integer>> result=new ArrayList<>();
        
        dfs(root,0,result);
        return result;
          
    }
    
    private void dfs(TreeNode root,int depth, List<List<Integer>> result)
    {
        if(root==null)
            return;
        if(result.size()==depth)
            result.add(new ArrayList<>());
        
        result.get(depth).add(root.val);
        
        dfs(root.left,depth+1,result);
        dfs(root.right,depth+1,result);
    }
    */
}